### ISSUE TITLE
topology: Extend the lifetime of IP–host ID mappings until they're not needed anymore

### ISSUE THREAD

**Issue by dawmd** [scylladb/scylladb#15968]
## Context
As of now, while decommissioning a node (or performing another topology operation similar in nature to it) we need to remove its tokens from token metadata. It's natural. However, it has consequences that can affect how Scylla operates significantly. One example of that is we cannot attempt to convert a node's IP to host ID (or vice versa) after that point. The reason for that is topology – the data structure that stores those mappings – is part of token metadata; when we remove the information about a node from it, its corresponding mappings disappear as well.

## Motivation
Said information is crucial in the transition from IPs to host IDs in Hinted Handoff as well as in storage proxy. Unless the conversion takes place simultaneously, we'll need to use the mappings at some point. Otherwise Scylla will stop operating properly until all necessary changes have been applied.

## Proposed solution
To tackle the problem, we should preserve the logic related to token metadata except for its part affecting its underlying topology. All information should be erased as it is now, but the topology should remain unchanged until the corresponding node has properly been decommissioned (/replaced, etc.). This way, mutations will still be routed to the right node and the token ring will operate as intended, but we'll still have access to the mappings.

## Problems
This solution brings a variety of challenges. Among them are:
* repair based operations rely on the state of topology. We can't let the information about a node remain there too long,
* Raft decommission and non-Raft decommission vary significantly. While modifying the latter shouldn't result in many changes (however significant they are), it's quite possible that modifying the former might pose a challenge.

## Another solution
If the approach above doesn't turn out feasible, we could try to extend token metadata or topology by a set of mappings corresponding to nodes that have been removed from token metadata recently. What the "recently" means in this context is to decide – we could either leave the mapping for a specified amount of time (very error-prone solution obviously. If decommissioning a node takes a significant amount of time, we might still end up removing the necessary information too early), or remove it once the topology change has properly finished.

This solution is arguably worse because it extends a data structure used throughout the system just for the sake of one use case. The advantage of it, however, is that it should be simpler.

**Comment by dawmd** [comment]
A short update: I'm currently trying to implement the alternative solution and test it. I artificially extend the information about the mappings corresponding to nodes affected by topology changes by keeping them as an additional field of token metadata. The places I modified are mostly in `storage_service.cc`: `node_ops_cmd_handler` (whenever `token_metadata` is mutated), `decommission`, `removenode`. When an operation starts, I store the relevant mapping, and when the *whole* operation finishes, only then do I remove it.
My testing is based on several asserts scattered throughout the hinted handoff module. Where I'd normally need to translate a host ID to the corresponding IP or the other way round, I check if the mapping exists. If it doesn't, Scylla should fail immediately. I tested my current work against three core d-tests: `hintedhandoff_additional_test`, `repair_based_node_operations_test` and `topology_test`. They all pass successfully. However, unit tests don't. There are two tests that sometimes pass, sometimes don't – `topology/test_change_ip` and `topology_experimental_raft/test_topology_ops`. I don't know what the problem is, but I'm working on it now.
So far I haven't done any more testing of this implementation. It seems some part of the test suite is flaky, so I'll wait with running any Jenkins jobs for the time being and focus on solving the local issues.
